Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -552,28 +552,32 @@ extension ActivityStreamPanel: DataObserverDelegate {
         }
         profile.history.removeHostFromTopSites(host).uponQueue(.main) { result in
             guard result.isSuccess else { return }
-            self.profile.panelDataObservers.activityStream.invalidate(highlights: false)
+            self.profile.panelDataObservers.activityStream.invalidateTopSites()
+            self.profile.panelDataObservers.activityStream.refresh(highlights: false)
         }
     }
 
     func pinTopSite(_ site: Site) {
         profile.history.addPinnedTopSite(site).uponQueue(.main) { result in
             guard result.isSuccess else { return }
-            self.profile.panelDataObservers.activityStream.invalidate(highlights: false)
+            self.profile.panelDataObservers.activityStream.invalidateTopSites()
+            self.profile.panelDataObservers.activityStream.refresh(highlights: false)
         }
     }
 
     func removePinTopSite(_ site: Site) {
         profile.history.removeFromPinnedTopSites(site).uponQueue(.main) { result in
             guard result.isSuccess else { return }
-            self.profile.panelDataObservers.activityStream.invalidate(highlights: false)
+            self.profile.panelDataObservers.activityStream.invalidateTopSites()
+            self.profile.panelDataObservers.activityStream.refresh(highlights: false)
         }
     }
 
     func hideFromHighlights(_ site: Site) {
         profile.recommendations.removeHighlightForURL(site.url).uponQueue(.main) { result in
             guard result.isSuccess else { return }
-            self.profile.panelDataObservers.activityStream.invalidate(highlights: true)
+            self.profile.panelDataObservers.activityStream.invalidateTopSites()
+            self.profile.panelDataObservers.activityStream.refresh(highlights: true)
```

Comment:
Why does this need to recompute Top Sites? Seems like you want an enum set here, so you can specify that you want to refresh Top Sites, Highlights, or both, no?